### PR TITLE
Scratchblocks v3

### PR DESCRIPTION
**Pull request by tjvr**
scratchblocks v3.1 isn't quite done yet, so this isn't ready to be merged yet.

I believe development on s2forums is now being done in private, and not in this repo; however, I figure a PR is still the easiest way to send you a patch!

### CODE REVIEW

**Comment by tjvr** [comment]
Here's a [diff showing code changes](https://github.com/tjvr/s2forums/compare/50be2cc081bd64df18f2b32067ddf3c740e5473a...tjvr:scratchblocks-v3).

**Comment by rschamp** [comment]
Thanks for your PR! Correct, this repo is here to accept PRs, but we've pulled this code into the private scratchr2 repo. When we merge the PRs here we then pull them in to scratchr2.

Does this include #1334 as well?  When this is ready to be merged, please let me know and I'll try it out.  Could you summarize the changes and put together a plan for how to test this?

**Comment by tjvr** [comment]
Hi! This is done. :) It just needs testing.

Changes:
- Rewrote renderer, now SVG-based (yay!)
- Rewrote parser (broke compatibility for a few edge cases and "hacks")
- Added "Add to backpack" button

Things to test:

- Browse to a discussion page (eg. <https://scratch.mit.edu/discuss/topic/14778/>) and check that blocks appear
- Browse to a post on a language forum (eg. German <https://scratch.mit.edu/discuss/post/307501/>), and check that most of the blocks are the right colour (nb. the plugin is now case-sensitive so a few old examples may no longer work)
- Browse to a post containing a (valid!) Scratch script (eg. <https://scratch.mit.edu/discuss/post/1942886/>) and click "Add to backpack"
- In the reply box, check the scratchblocks menu still works (screenshot below) -- the menu should display, and clicking blocks should insert them (it's pretty messy--I really don't like markitup--so you can't ask for more than that :P)
- Hit preview, and check that blocks are rendered inside the preview too
- Also test the menu on a language forum topic
- The preview has historically been buggy on language forum topics (the preview has no way to detect what language forum you're on, so it can only render blocks in English), but that might be worth a look too

Hope this helps. Sorry, this looks like a lot—but it should take less than five minutes!

screen shot 2016-04-24 at 22 32 52

**Comment by tjvr** [comment]
I'm gonna keep this updated with scratchblocks master, but this is good to be merged anytime (assuming it works!) :)

**Comment by tjvr** [comment]
I have finally found a way to test this locally (using mitmproxy to modify the HTTPS flow). I've found some bugs; I'll update this PR and let you know when I'm happy. :)

**Comment by tjvr** [comment]
Right, I'm happy with this. Hopefully it should all work! :)

**Comment by rschamp** [comment]
Ok, thanks! I'm going to make the PR on the internal repo now so that we can try this out.

**Comment by tjvr** [comment]
Thank you! Let me know how you get on :)

**Comment by tjvr** [comment]
@rschamp Did this work okay? I just discovered (and fixed!) a slow rendering bug—see <https://github.com/tjvr/scratchblocks/issues/167>—should I add that patch here?

**Comment by rschamp** [comment]
Thanks, yes please update that here. Sorry I didn't get around to reviewing this for this release, but I'll look at it next week!

**Comment by thisandagain** [comment]
@rschamp Bump

**Comment by edward789121** [comment]
> Here's a [diff showing code changes.](https://github.com/tjvr/s2forums/compare/50be2cc081bd64df18f2b32067ddf3c740e5473a...tjvr:scratchblocks-v3)

or check it out [here](https://github.com/edward789121/s2forums).

**Comment by hyperobject** [comment]
@rschamp Should this be moved to the July milestone?

**Comment by rschamp** [comment]
Thanks for taking this over @technoboy10

**Comment by tjvr** [comment]
@technoboy10 nooo do it today

File: djangobb_forum/static/djangobb_forum/scratchblocks/backpack.js [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+    var USER = window.Scratch && window.Scratch.INIT_DATA.LOGGED_IN_USER.model?
+           window.Scratch.INIT_DATA.LOGGED_IN_USER.model.username :
+           null;
+    if (!(CSRFTOKEN && USER)) {
+      return no('Not logged in.');
+    }
+    var x = new XMLHttpRequest();
+    x.open(
+      "GET",
+      "//scratch.mit.edu/internalapi/backpack/"+USER+"/get/",
```
Comment (hyperobject) [inline djangobb_forum/static/djangobb_forum/scratchblocks/backpack.js:39]: This URL, along with the one on line 52, should be relative. Fixing downstream. :)

**Comment by hyperobject** [comment]
@tjvr @Hardmath123 On posts with scripts that are longer than the width of a post (e.g. [this post](https://scratch.mit.edu/discuss/topic/66190/?page=1#post-561900)), the "Add to Backpack" button doesn't respond to clicks.

screenshot 2016-06-21 14 17 53

**Comment by hyperobject** [comment]
Once that's fixed, this should be good to merge.

File: djangobb_forum/static/djangobb_forum/scratchblocks/backpack.js (hunk 1 above)
Comment (tjvr) [inline djangobb_forum/static/djangobb_forum/scratchblocks/backpack.js:39]: Thanks you're the best

**Comment by tjvr** [comment]
@technoboy10 I think this should fix that but mitmproxy is making me cry so who knows?

**Comment by hyperobject** [comment]
@tjvr @Hardmath123 Thanks guys, this is awesome! I've merged this in the internal repo.
